restapi/src/main/java/io/stargate/web/docsapi/service/JsonConverterService.java

```diff
@@ -0,0 +1,290 @@
+  }
+
+  public ImmutablePair<JsonNode, Map<String, List<JsonNode>>> convertToJsonDoc(
+      List<Row> rows, boolean writeAllPathsAsObjects, boolean numericBooleans) {
+    JsonNode doc = mapper.createObjectNode();
+    Map<String, Long> pathWriteTimes = new HashMap<>();
+    Map<String, List<JsonNode>> deadLeaves = new HashMap<>();
+    if (rows.isEmpty()) {
+      return ImmutablePair.of(doc, deadLeaves);
+    }
+    Column writeTimeCol = Column.reference("writetime(leaf)");
+
+    for (Row row : rows) {
+      Long rowWriteTime = row.getLong(writeTimeCol.name());
+      String rowLeaf = row.getString("leaf");
+      if (rowLeaf.equals(DocumentDB.ROOT_DOC_MARKER)) {
+        continue;
+      }
+
+      String leaf = null;
+      JsonNode parentRef = null;
+      JsonNode ref = doc;
+
+      String parentPath = "$";
+
+      for (int i = 0; i < DocumentDB.MAX_DEPTH; i++) {
+        String p = row.getString("p" + i);
+        String nextP = i < DocumentDB.MAX_DEPTH - 1 ? row.getString("p" + (i + 1)) : "";
+        boolean endOfPath = nextP.equals("");
+        boolean isArray = p.startsWith("[");
+        boolean nextIsArray = nextP.startsWith("[");
+
+        if (isArray) {
+          // This removes leading zeros if applicable
+          p = "[" + Integer.parseInt(p.substring(1, p.length() - 1)) + "]";
+        }
+
+        boolean shouldWrite =
+            !pathWriteTimes.containsKey(parentPath)
+                || pathWriteTimes.get(parentPath) <= rowWriteTime;
+
+        if (!shouldWrite) {
+          markFullPathAsDead(parentPath, p, deadLeaves);
+          break;
+        }
+
+        if (endOfPath) {
+          boolean shouldBeArray = isArray && !ref.isArray() && !writeAllPathsAsObjects;
+          if (i == 0 && shouldBeArray) {
+            doc = mapper.createArrayNode();
+            ref = doc;
+            pathWriteTimes.put(parentPath, rowWriteTime);
+          } else if (i != 0 && shouldBeArray) {
+            markObjectAtPathAsDead(ref, parentPath, deadLeaves);
+            ref = changeCurrentNodeToArray(row, parentRef, i);
+            pathWriteTimes.put(parentPath, rowWriteTime);
+          } else if (i != 0 && !isArray && !ref.isObject()) {
+            markArrayAtPathAsDead(ref, parentPath, deadLeaves);
+            ref = changeCurrentNodeToObject(row, parentRef, i, writeAllPathsAsObjects);
+            pathWriteTimes.put(parentPath, rowWriteTime);
+          }
+          leaf = p;
+          break;
+        }
+
+        JsonNode childRef;
+
+        if (isArray && !writeAllPathsAsObjects) {
+          if (!ref.isArray()) {
+            if (i == 0) {
+              doc = mapper.createArrayNode();
+              ref = doc;
+              pathWriteTimes.put(parentPath, rowWriteTime);
+            } else {
+              markObjectAtPathAsDead(ref, parentPath, deadLeaves);
+              ref = changeCurrentNodeToArray(row, parentRef, i);
+              pathWriteTimes.put(parentPath, rowWriteTime);
```

Comment:
`pathWriteTimes.put(parentPath, rowWriteTime);`  at lines 102 and 106 could be extracted out of the inner if else (`if (i==0)`)